CodingTracker.TildaDares/Infrastructure/CodingTrackerDatabase.cs

```diff
@@ -0,0 +1,237 @@
+using System.Configuration;
+using CodingTracker.TildaDares.Models;
+using Dapper;
+using Microsoft.Data.Sqlite;
+using Spectre.Console;
+
+namespace CodingTracker.TildaDares.Infrastructure;
+
+public class CodingTrackerDatabase
+{
+    private readonly string _connectionString = ConfigurationManager.ConnectionStrings["CodingTrackerDB"].ConnectionString;
+    
+    public CodingTrackerDatabase()
+    {
+        CreateCodingTrackerDb();
+    }
+
+    public void InsertCodingSession(CodingSession codingSession)
+    {
+        using var connection = new SqliteConnection(_connectionString);
+        try
+        {
+            connection.Open();
+            const string sql = "INSERT INTO codingTracker(startTime, endTime, duration) VALUES (@StartTime, @EndTime, @Duration)";
+            var rowsAffected = connection.Execute(sql, codingSession);
+            AnsiConsole.MarkupLine($"[green]{rowsAffected} row(s) inserted.[/]");
+        }
+        catch (SqliteException e)
+        {
+            AnsiConsole.MarkupLine($"[red]Unable to insert into database. {e.Message}[/]");
+        }
+        finally
+        {
+            connection.Close();
+        }
+    }
+
+    public CodingSession GetCodingSession(int id)
+    {
+        using var connection = new SqliteConnection(_connectionString);
+        CodingSession session = null;
+        try
+        {
+            connection.Open();
+            const string sql = "SELECT * FROM codingTracker WHERE id = @Id";
+            session = connection.QuerySingleOrDefault<CodingSession>(sql, new {Id = id});
+        }
+        catch (SqliteException e)
+        {
+            AnsiConsole.MarkupLine($"[red]Unable to retrieve coding session record with ID: {id}. {e.Message}[/]");
+        }
+        finally
+        {
+            connection.Close();
+        }
+        
+        return session;
+    }
+
+    public CodingStats GetSumOfCodingSessionDuration(CodingSessionFilter filter)
+    {
+        using var connection = new SqliteConnection(_connectionString);
+        CodingStats stats = null;
+        try
+        {
+            connection.Open();
+            var sql = "SELECT printf('%.2f', SUM(duration)) AS TotalHours, COUNT(*) As RecordCount FROM codingTracker";
+            if (filter is { StartTime: not null, EndTime: not null })
+            {
+                sql += " WHERE DATETIME(startTime) >= DATETIME(@StartTime) AND DATETIME(endTime) <= DATETIME(@EndTime)";
```

Comment:
🟠 While one/two string concatenations are fine.

🔧 I would suggest transitioning to a string builder or query builder object to manage any performance impacts. (although negligible in these learning apps.)

Reply:
I'll take note of this in future projects!